spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/data/vectorized/comet/CometIcebergColumnarBatchReader.java

```diff
@@ -0,0 +1,303 @@
+      this.numRowsToRead = numRowsToRead;
+      if (hasIsDeletedColumn) {
+        isDeleted = new boolean[numRowsToRead];
+      }
+    }
+
+    ColumnarBatch loadDataToColumnBatch() {
+      int numRowsUndeleted = initRowIdMapping();
+
+      ColumnVector[] arrowColumnVectors = readDataToColumnVectors();
+
+      ColumnarBatch newColumnarBatch = new ColumnarBatch(arrowColumnVectors);
+      newColumnarBatch.setNumRows(numRowsUndeleted);
+
+      if (hasEqDeletes()) {
+        applyEqDelete(newColumnarBatch);
+      }
+
+      if (hasIsDeletedColumn && rowIdMapping != null) {
+        // reset the row id mapping array, so that it doesn't filter out the deleted rows
+        for (int i = 0; i < numRowsToRead; i++) {
+          rowIdMapping[i] = i;
+        }
+        newColumnarBatch.setNumRows(numRowsToRead);
+      }
+
+      if (hasIsDeletedColumn) {
+        readDeletedColumnIfNecessary(arrowColumnVectors);
+      }
+
+      return newColumnarBatch;
+    }
+
+    ColumnVector[] readDataToColumnVectors() {
+      ColumnVector[] columnVectors = new ColumnVector[readers.length];
+      // Fetch rows for all readers in the delegate
+      delegate.nextBatch(numRowsToRead);
+      for (int i = 0; i < readers.length; i++) {
+        CometIcebergVector bv = ((CometIcebergColumnReader) readers[i]).getVector();
+        bv.setDelegate(((CometIcebergColumnReader) readers[i]).getDelegate().currentBatch());
+        bv.setRowIdMapping(rowIdMapping);
+        columnVectors[i] = bv;
+      }
+
+      return columnVectors;
+    }
+
+    boolean hasEqDeletes() {
+      return deletes != null && deletes.hasEqDeletes();
+    }
+
+    int initRowIdMapping() {
+      Pair<int[], Integer> posDeleteRowIdMapping = posDelRowIdMapping();
+      if (posDeleteRowIdMapping != null) {
+        rowIdMapping = posDeleteRowIdMapping.first();
+        return posDeleteRowIdMapping.second();
+      } else {
+        rowIdMapping = initEqDeleteRowIdMapping();
+        return numRowsToRead;
+      }
+    }
+
+    Pair<int[], Integer> posDelRowIdMapping() {
+      if (deletes != null && deletes.hasPosDeletes()) {
+        return buildPosDelRowIdMapping(deletes.deletedRowPositions());
+      } else {
+        return null;
+      }
+    }
+
+    /**
+     * Build a row id mapping inside a batch, which skips deleted rows. Here is an example of how we
+     * delete 2 rows in a batch with 8 rows in total. [0,1,2,3,4,5,6,7] -- Original status of the
+     * row id mapping array [F,F,F,F,F,F,F,F] -- Original status of the isDeleted array Position
+     * delete 2, 6 [0,1,3,4,5,7,-,-] -- After applying position deletes [Set Num records to 6]
+     * [F,F,T,F,F,F,T,F] -- After applying position deletes
+     *
+     * @param deletedRowPositions a set of deleted row positions
+     * @return the mapping array and the new num of rows in a batch, null if no row is deleted
+     */
+    Pair<int[], Integer> buildPosDelRowIdMapping(PositionDeleteIndex deletedRowPositions) {
```

Comment:
Same question about the overlap here.

Reply:
Will reuse the common code.

Reply:
Please check if a common parent or a util class would be better.